File: codes/perl/inheritance/multiple/multiple2.pl
package Person;

sub new {
    my $class = shift;
    my $self = {
        name => shift,
        serial => 5
    };
    bless $self, $class;
    return $self;
}

sub info {
    my ($self) = @_;
    print "Person name: $self->{name}\n";
    print "Person serial: $self->{serial}\n";
}

1;

package Machine;

sub new {
    my $class = shift;
    my $self = {serial => shift};
    bless $self, $class;
    return $self;
}

sub info {
    my ($self) = @_;
    print "Machine serial: $self->{serial}\n";
}

1;

package Android;
our @ISA = qw(Machine Person);

sub new {
    my ($class, $name, $serial) = @_;
    my $self = $class->Person::new($name);
    my $self2 = $class->Machine::new($serial);
    $self = { %$self, %$self2};
    bless $self, $class;
    return $self;
}

1;

package Cyborg;
our @ISA = qw(Person Machine);

sub new {
    my ($class, $name, $serial) = @_;
    my $self = $class->Machine::new($serial);
    my $self2 = $class->Person::new($name);
    $self = { %$self, %$self2};
    bless $self, $class;
    return $self;
}

1;

$p = Person->new("person");
$p->info();
print "=====================\n";
$m = Machine->new(15);
$m->info();
print "=====================\n";
$a = Android->new("android", 17);
$a->info();
print "=====================\n";
$c = Cyborg->new("cyborg", 20);
$c->info();
print "=====================\n";

File: codes/php/inheritance/multiple/multiple2.php
<?php

interface iPerson {
    public function info();
}

interface iMachine {
    public function info();
}

trait tPerson {
    public function info() {
        echo "My name: $this->name\n";
    }
}

trait tMachine {
    public function info() {
        echo "My serial number: $this->serial\n";
    }
}

class Person implements iPerson {
    use tPerson;
    private $name;
    public function __construct($name) {
        $this->name = $name;
    }
}

class Machine implements iMachine {
    use tMachine;
    private $serial;
    public function __construct($serial) {
        $this->serial = $serial;
    }
}

class Android extends Machine implements iPerson {
    use tPerson { info as private pinfo;}
    private $name;

    public function __construct($name, $serial) {
        parent::__construct($serial);
        $this->name = $name;
    }

    public function info() {
        parent::info();
        $this->pinfo();
    }

}

class Cyborg extends Person implements iMachine {
    use tMachine { info as private minfo;}
    private $serial;

    public function __construct($name, $serial) {
        parent::__construct($name);
        $this->serial = $serial;
    }

    public function info() {
        parent::info();
        $this->minfo();
    }

}

$p = new Person("person1");
$p->info();
echo "=====================\n";
$m = new Machine(15);
$m->info();
echo "=====================\n";
$a = new Android("android1", 16);
$a->info();
echo "=====================\n";
$c = new Cyborg("cyborg1", 17);
$c->info();
echo "=====================\n";

?>
